lnwallet/chanfunding/coin_select.go

```diff
@@ -251,3 +254,68 @@ func CoinSelectSubtractFees(feeRate chainfee.SatPerKWeight, amt,
 
 	return selectedUtxos, outputAmt, changeAmt, nil
 }
+
+// CoinSelectUpToAmount attempts to select coins such that we'll select up to
+// maxAmount exclusive fees if sufficient funds are available. If insufficient
+// funds are available this method selects all available coins.
+func CoinSelectUpToAmount(feeRate chainfee.SatPerKWeight, maxAmount,
+	dustLimit btcutil.Amount, coins []Coin) ([]Coin, btcutil.Amount,
+	btcutil.Amount, error) {
+
+	// Get balance from coins.
+	var totalBalance btcutil.Amount
+	for _, coin := range coins {
+		totalBalance += btcutil.Amount(coin.Value)
+	}
+
+	// First we try to select coins to create an output of the specified
+	// maxAmount and a change output that covers the miner fee.
+	selectedCoins, changeAmt, err := CoinSelect(
+		feeRate, maxAmount, dustLimit, coins,
+	)
+
+	outputAmount := maxAmount
+
+	// Track if our coin selection was unsuccessful and whether we have to
+	// start a new round of selecting coins considering fees.
+	selectSubtractFee := false
+
+	var insufficientFundsError *ErrInsufficientFunds
+	switch {
+	case err == nil:
+	// If the initial coin selection fails due to insufficient funds we
+	// select our total available balance minus fees.
+
+	case errors.As(err, &insufficientFundsError):
+		selectSubtractFee = true
+
+	default:
+		return nil, 0, 0, err
+	}
+
+	// If we determined that our local balance is insufficient we check
+	// our total balance minus fees.
+	if selectSubtractFee {
+		// nolint:lll
+		selectedCoins, outputAmount, changeAmt, err = CoinSelectSubtractFees(
+			feeRate, totalBalance, dustLimit, coins,
+		)
+		if err != nil {
+			return nil, 0, 0, err
+		}
+	}
+
+	// Sanity check: Always make sure the output amount is below the
+	// specified max amount.
+	switch overshoot := outputAmount - maxAmount; {
+	case overshoot > dustLimit:
+		return nil, 0, 0, fmt.Errorf("non-negative difference in "+
+			"requested maximum funding amount %v and obtained "+
+			"amount %v", maxAmount, outputAmount)
+
+	case overshoot > 0:
```

Comment:
Good catch! This scenario could have arisen before in the interplay of `CoinSelect` and `CoinSelectSubtractFees` as described originally here: https://github.com/lightningnetwork/lnd/pull/4029#issuecomment-592955663. Basically if the `CoinSelect` function threw an insufficient funds error expecting sufficient fee for one change output but the available balance would have been 1 sat + `fundMaxAmt`+ fee for one input and no change output, then the `CoinSelectSubtractFee` could have selected more than `fundMaxAmt` since its calculating fees without a change output.

But since `CoinSelect` was changed to also consider coin selections without change output I think this check can be removed.

Do you agree?